snakemake/executors.py

```diff
@@ -1492,6 +1492,10 @@ def run(self, job, callback=None, submit_callback=None, error_callback=None):
             container.resources.requests["memory"] = "{}M".format(
                 job.resources["mem_mb"]
             )
+        if "nvidia_gpu" in job.resources.keys():
+            container.resources.limits["nvidia/gpu"] = "%d".format(
```

Comment:
I like this, but it needs to be generic. Is there a fixed format for resources names in kubernetes? I.e. always separated by a slash?

Reply:
Maybe you already pick up on this feature, about naming gpu resources, it depends on GPU vendor. I'm not certain but just saying, this rule started with nvidia. They just named on resource with `nvidia.com/gpu`. And many cloud vendor is now supporting nvidia-gpu-enabled VM. After time passed, recently AMD is also named their gpu resource as `amd.com/gpu`

You can check kubernetes pod manifests with below links.

* https://github.com/NVIDIA/k8s-device-plugin#running-gpu-jobs
* https://github.com/RadeonOpenCompute/k8s-device-plugin/blob/master/example/pod/alexnet-gpu.yaml

Actually, my previous commit has problem because of missing `.com`, I'll fix it.

Reply:
Ok. That side is good then. So you could also add handling for `amd_gpu`. Does a generic `gpu` make sense? I guess only if it would be possible to query available gpus beforehand via the kubernetes api?

Reply:
@joonjeong can you add amd_gpu handling as well?
@SooLee is there a way to pass gpu requirements via tibanna?
@vsoch  is there is a way to pass gpu requirements via GLS?

Reply:
On it! For a brief look over, it looks like the user can select an N1 instance type and then an accelerator type and count. So we could expose that to the user by way of a --gpu N flag (or similar) that would then select an instance from the N1 family and configure an accelerator. How is the current UI (command line or snakefile) setup to receive a gpu request? Is it part of the job metadata?

@johanneskoester I'll have an email going out to you and Andrew shortly with some questions, and once I'm familiar with how the input is done with snakemake, I can add to the Google Life Sciences PR. Stay tuned!

Reply:
> On it! For a brief look over, it looks like the user can select an N1 instance type and then an accelerator type and count. So we could expose that to the user by way of a --gpu N flag (or similar) that would then select an instance from the N1 family and configure an accelerator. How is the current UI (command line or snakefile) setup to receive a gpu request? Is it part of the job metadata?
> 
> @johanneskoester I'll have an email going out to you and Andrew shortly with some questions, and once I'm familiar with how the input is done with snakemake, I can add to the Google Life Sciences PR. Stay tuned!

People will define the needed gpu resource in the resources definition of the rule:
```
rule a:
    output:
        "test.txt"
    resources:
        nvidia_gpu=1
    shell:
        "somecommand ..."
```

We might expose more control in the future, but that's it for now.

Reply:
okay, on it! I think GCP might just have support for nvidia, so we wouldn't have the middle option. I'll dig more into it when I code this up.

Reply:
Just so we have it later, here are the acceleratorTypes that are available for an n1 (reasonably sized) instance (it's based on zone actually):

```
{'id': 'projects/snakemake-testing/zones/us-west1-a/acceleratorTypes',
 'items': [{'id': '10004',
   'creationTimestamp': '1969-12-31T16:00:00.000-08:00',
   'name': 'nvidia-tesla-p100',
   'description': 'NVIDIA Tesla P100',
   'zone': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a',
   'selfLink': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a/acceleratorTypes/nvidia-tesla-p100',
   'maximumCardsPerInstance': 4,
   'kind': 'compute#acceleratorType'},
  {'id': '10007',
   'creationTimestamp': '1969-12-31T16:00:00.000-08:00',
   'name': 'nvidia-tesla-p100-vws',
   'description': 'NVIDIA Tesla P100 Virtual Workstation',
   'zone': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a',
   'selfLink': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a/acceleratorTypes/nvidia-tesla-p100-vws',
   'maximumCardsPerInstance': 4,
   'kind': 'compute#acceleratorType'},
  {'id': '10019',
   'creationTimestamp': '1969-12-31T16:00:00.000-08:00',
   'name': 'nvidia-tesla-t4',
   'description': 'NVIDIA Tesla T4',
   'zone': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a',
   'selfLink': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a/acceleratorTypes/nvidia-tesla-t4',
   'maximumCardsPerInstance': 4,
   'kind': 'compute#acceleratorType'},
  {'id': '10020',
   'creationTimestamp': '1969-12-31T16:00:00.000-08:00',
   'name': 'nvidia-tesla-t4-vws',
   'description': 'NVIDIA Tesla T4 Virtual Workstation',
   'zone': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a',
   'selfLink': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a/acceleratorTypes/nvidia-tesla-t4-vws',
   'maximumCardsPerInstance': 4,
   'kind': 'compute#acceleratorType'},
  {'id': '10008',
   'creationTimestamp': '1969-12-31T16:00:00.000-08:00',
   'name': 'nvidia-tesla-v100',
   'description': 'NVIDIA Tesla V100',
   'zone': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a',
   'selfLink': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a/acceleratorTypes/nvidia-tesla-v100',
   'maximumCardsPerInstance': 8,
   'kind': 'compute#acceleratorType'}],
 'selfLink': 'https://www.googleapis.com/compute/v1/projects/snakemake-testing/zones/us-west1-a/acceleratorTypes',
 'kind': 'compute#acceleratorTypeList'}
```
I'm thinking we want the smallest number that is greater than or equal to the number of gpus requested.

Reply:
I'm not familiar with the subtle differences between these, and if we would need to add some ability to specify, for example, the T4 vs P100. For now I'll chose the T4 since it seems slightly better, and we can add some kind of variable if a user requests it in the future. https://www.microway.com/hpc-tech-tips/nvidia-turing-tesla-t4-hpc-performance-benchmarks/ like, damn! GPUs...!

Reply:
okay all set! https://github.com/snakemake/snakemake/pull/90#issuecomment-611173810

Reply:
Thank you guys for reviewing my opinion!

@johanneskoester Ok, I'll attach more commit on this branch.

Reply:
One thing I'd like to tell more, there's no way to select GPU type in kubernetes manifest because it's hard to define generic way, so nvidia-docker solve this problem with [OCI(Open Container Initiative) Runtime Spec](https://github.com/opencontainers/runtime-spec) and you guys can find the way how to configure GPU options on container in [this article](https://devblogs.nvidia.com/gpu-containers-runtime/) and you can check how it's implemented in  [nvidia-docker](https://github.com/NVIDIA/nvidia-docker) repository.

I tested this method on AWS and Google Cloud but to describe how it works, there should be more examples. I'll prepare more about this example soon.

Reply:
PR #329 introduces string resources. In short, people could specify `gpu_model="nvidia-tesla-p100`. Would that be feasible for kubernetes as well @joonjeong?

Reply:
@johanneskoester That could be good for KubernetesExecutor to enforce snakemake pod to select specific gpu model. but `nvidia.com/gpu` still needed because kubernetes decide to allocate actual resource(gpu) with specified under `resources` section.

So, if `gpu_model` keyword is added, I guess command line will be written like this.

```bash
$ snakemake --kubernetes --resource gpu_model=nvidia-tesla-p100 nvidia_gpu=1
```

Anyway, I think it's good to add `machine_type` and `gpu_model`. I'll also take care about that in this commit.

To @johanneskoester @vsoch, resource keywords are decided? I'll follow that keywords

Reply:
@joonjeong we had discussed `gpu_model` and `machine` but I like your suggestion of `machine_type` to be more specific, and I can change my PR if that is @johanneskoester preference.

Reply:
> @johanneskoester That could be good for KubernetesExecutor to enforce snakemake pod to select specific gpu model. but `nvidia.com/gpu` still needed because kubernetes decide to allocate actual resource(gpu) with specified under `resources` section.
> 
> So, if `gpu_model` keyword is added, I guess command line will be written like this.
> 
> ```shell
> $ snakemake --kubernetes --resource gpu_model=nvidia-tesla-p100 nvidia_gpu=1
> ```

I think nvidia_gpu can be omitted, because the KubernetesExecutor can simply infer that from the gpu_model definition, right?

@vsoch @joonjeong `machine_type` sounds good.

Reply:
@johanneskoester all set, updated machine to machine_type https://github.com/snakemake/snakemake/pull/90/commits/81a5b2507227c969e6d68473370a6a0c77afc4df